### PR TITLE
LF: runtime support for BigNumeric

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This is part of #8719

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -852,6 +853,90 @@ private[lf] object SBuiltin {
...
+      val y = args.get(1).asInstanceOf[SBigNumeric].value
+      rightOrArithmeticError("overflow/underflow", SBigNumeric.fromBigDecimal(x multiply y))
+    }
+  }
+
+  final object SBDivBigNumeric extends SBuiltinPure(4) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SBigNumeric = {
+      val scale = rightOrCrash(SBigNumeric.checkScale(args.get(0).asInstanceOf[SInt64].value))
+      val roundingMode =
+        java.math.RoundingMode.valueOf(args.get(1).asInstanceOf[SInt64].value.toInt)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:898]: What happens for invalid scales? Does this crash or default to something weird?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 2]
```diff
@@ -210,7 +214,25 @@ object SValue {
   // with SValue and we can remove one layer of indirection.
   sealed trait SPrimLit extends SValue with Equals
   final case class SInt64(value: Long) extends SPrimLit
+  // TODO https://github.com/digital-asset/daml/issues/8719
+  //  try to factorize SNumeric and SBigNumeric
+  //  not it seems that scale is relevant in SNumeric but lost in SBigNumeric
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:219]: typo? Not quite sure what the `not` is supposed to mean here.

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -315,6 +315,7 @@ private[lf] object SBuiltin {
         case SParty(p) => p
         case SUnit => s"<unit>"
         case SDate(date) => date.toString
+        case SBigNumeric(x) => Numeric.toString(x)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:318]: `x` is always without trailing.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -852,6 +853,90 @@ private[lf] object SBuiltin {
     }
   }
 
+  final object SBScaleBigNumeric extends SBuiltinPure(1) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SInt64 = {
+      val x = args.get(0).asInstanceOf[SBigNumeric].value
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:858]: No real policy. If you prefer pattern matching, I will change it.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:898]: If scale is invalid it crash.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 2 above)
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:219]:
```suggestion
  //  note it seems that scale is relevant in SNumeric but lost in SBigNumeric
```

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: "note".

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:858]: I don’t (in this instance), I’m perfectly fine keeping it. I was more curious if I should keep something in mind for when I touch this next.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:318]:
```
  final def toString(x: BigDecimal): String = {
    val s = x.toPlainString
    if (x.scale <= 0) s + "." else s
```

**Review by sofiafaro-da — APPROVED** [review]
Thanks :-)

**Pull request merged**